src/System.Net.Http/src/System/Net/Http/Managed/HttpEnvironmentProxy.cs

```diff
@@ -0,0 +1,199 @@
+            if (value != null)
+            {
+                http = value;
+            }
+            value = Environment.GetEnvironmentVariable("HTTPS_PROXY");
+            if (value != null)
+            {
+                https = value;
+            }
+
+            value = Environment.GetEnvironmentVariable("https_proxy");
+            if (value != null)
+            {
+                https = value;
+            }
+            // fail to instantiate if nothing is set.
+            // caller may pick some other proxy type.
+            if (String.IsNullOrWhiteSpace(http) && String.IsNullOrWhiteSpace(https))
+            {
+                return null;
+            }
+
+            return new HttpEnvironmentProxy(http, https);
+        }
+
+        private HttpEnvironmentProxy(string http, string https)
+        {
+
+            if (!String.IsNullOrWhiteSpace(http))
+            {
+                _http = GetUriFromString(http);
+                if (!string.IsNullOrWhiteSpace(_http.UserInfo))
+                {
+                    String[] s = _http.UserInfo.Split(':', 2);
+                    _httpCred = new NetworkCredential(s[0], s[1]);
+                }
+            }
+            if (!String.IsNullOrWhiteSpace(https))
+            {
+                _https = GetUriFromString(https);
+                if (!string.IsNullOrWhiteSpace(_https.UserInfo))
+                {
+                    String[] s = _https.UserInfo.Split(':', 2);
+                    _httpsCred = new NetworkCredential(s[0], s[1]);
+                }
+            }
+
+            string value = Environment.GetEnvironmentVariable("no_proxy");
+            if (!String.IsNullOrWhiteSpace(value)) 
+            {
+                _bypass = value.Split(',');
+            }
+        }
+
+        /// <summary>
+        /// This function will evaluate given string and it will try to convert
+        /// it to Uri object. The string could contain URI fragment, IP address and  port
+        /// tuple or just IP address or name. 
+        /// </summary>
+        private Uri GetUriFromString(String value)
+        {
+            if (!value.StartsWith("http://", StringComparison.OrdinalIgnoreCase) &&
+                !value.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
+            {
+                // Default to HTTP
+                value = "http://" + value;
+            }
+            Uri uri = new Uri(value);
+
+            if (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps)
+            {
+                throw new NotSupportedException("Only HTTP and HTTPS protocols are supported for proxy.");
+            }
+            return uri;
+        }
+
+        /// <summary>
+        /// This function returns true if given Host mach bypss list.
+        /// Note, that the list is common for http and https.
+        /// </summary>
+        private bool IsMatchInBypassList(Uri input)
+        {
```

Comment:
Looks like _bypass is needed only when this function is called, can we move the instantiation of _bypass from ```HttpEnvironmentProxy``` method to here?

Reply:
I think it's actually ok where it is.  By doing the work in the ctor, which happens once per handler, we don't have to do any additional work (even checking whether it's been parsed) here, which happens potentially per request.